Delegation support for libcephfs

jtlayton opened the pull request:
This is a revised patchset to add delegation support to libcephfs.

It starts with a small, unrelated cleanup patch for common_preinit and global_pre_init.

From there, I reworked the set to do much less pointer and list handling, and changed it over to use exceptions to catch allocation failures.


I've also reworked the interface to use its own set of command constants for requesting delegated state. Along the way, I squashed most of the patches down into a smaller number of patches -- it was too hard to do the rework and keep the patches small.

Since ceph doesn't naturally break delegations on open, this patchset adds a tunable that allows the client to request that behaviour, in addition to setting the delegation timeout. This is implemented by having the client explicitly wait on some caps after the open call.

The tunables also use the observer to watch for config changes now, which is more efficient than calling get_val on every delegation break and open (for instance).

Finally, the testcases have been expanded to cover link, unlink and rename events as well as opens.

batrick commented:
Old PR for posterity: https://github.com/ceph/ceph/pull/17144

gregsfortytwo commented on src/client/Client.h:
```diff
@@ -404,6 +408,8 @@ class Client : public Dispatcher, public md_config_obs_t {
   entity_name_t get_myname() { return messenger->get_myname(); } 
   void _sync_write_commit(Inode *in);
+  void wait_on_list(list<Cond*>& ls);
+  void signal_cond_list(list<Cond*>& ls);
```
You could keep this private by declaring Delegation (or Fh, I forget which one is invoking it) as a friend. Not quite sure which is preferable...

jtlayton replied:
This seems like a pretty innocuous set of methods to expose.

gregsfortytwo replied:
Nah, that's fine, it just stuck out a bit in the diff and I had to think through what was going on.

jtlayton commented on src/include/cephfs/libcephfs.h:
```diff
@@ -1572,6 +1572,51 @@ int ceph_ll_getlk(struct ceph_mount_info *cmount,
+ *
+ * The application can set the client_deleg_timeout config option to suit its
+ * needs, but it should take care to choose a value that allows it to avoid
+ * forcible eviction from the cluster in the event of an application bug.
```
Ok, I'll need some guidance there. IIUC, the eviction timeout is supposed to be settable via a conf option, so in principle we should be able to just use ceph_conf_get to get the current value.

@jcsp : do you know how we would go about querying for this?

gregsfortytwo replied:
MDSMap::get_session_timeout(). It's distributed in the MDSMap so that we know clients and MDSes aren't disagreeing on the value.
(Unfortunately, it's just a bare variable rather than wrapped up in some extensible config options thing we could use elsewhere, but such is life...)

jtlayton replied:
Hmm ok. I'm not sure I like the idea of shutting down on that -- it'd be nice if we could just reject changing that value to anything less than get_session_timeout returns.  The config observer doesn't really allow that -- the callbacks there are all void return.

The next best thing would be to just ignore any attempt to set it to an unreasonable value. Basically, limit the minimum timeout to what get_session_timeout returns, and probably log a message if we had to alter it.

The tricky bit with all of this is that we may not have the mdsmap yet when we start parsing the config. So, what we may need to do is:

1) have the handle_conf_change routine call get_session_timeout. If it's !=0 then we'll use that to cap the value of the conf change. If we end up altering it, we log a warning.

2) if get_session_timeout returns 0, then we don't know the timeout yet. Just accept what userland gave us.

3) when we get a new map (in Client::handle_mds_map), get the session timeout and vet it against the delegation timeout. Reduce the deleg timeout if necessary, and log a message if that occurs.

gregsfortytwo replied:
I know you're writing all the NFS Ganesha code, but it seems like we really want a negotiation happening between Client and the CephFS FSAL so that both sides agree things are kosher before they proceed. If CephFS is silently setting a lower recall timeout, then Ganesha is going to be getting itself unexpectedly blacklisted, and that just sounds like a nightmare to debug.

Maybe discuss with @ajarr and the Manila people what they'd prefer to have happen in the case of misconfiguration?

jtlayton commented on src/client/Inode.h:
```diff
@@ -291,6 +297,34 @@ struct Inode {
+		  open_by_mode[CEPH_FILE_MODE_WR]);
+  };
+
+  // how many opens currently vs. this inode
```
Yes, that's what I meant. I'll reword it.

jtlayton commented on src/common/options.cc:
```diff
@@ -5981,6 +5981,14 @@ std::vector<Option> get_mds_client_options() {
+    .set_description("time (in seconds) that application has to return delegation after recall"),
+
+    Option("client_deleg_break_on_open", Option::TYPE_BOOL, Option::LEVEL_ADVANCED)
+    .set_default(false)
```
My thinking was that the application (ganesha and probably samba) could just call:

ceph_conf_set("client_deleg_break_on_open", "true");

whenever it creates a client. Since this is all about what sort of behaviour the application wants, I think that the simplest approach that doesn't change, e.g. ceph-fuse.

gregsfortytwo replied:
So what happens if we have a native CephFS client that doesn't know about these? Is it okay for that to open a file without delegations being broken, as long as it doesn't do any reads or writes?
I thought we needed to be able to actually deny opens until delegations are broken. And that means every client in the cluster, regardless of if they know about NFS directly, needs to have the same behavior.

jtlayton replied:
It really comes down to applications that need to enforce these sort of semantics. We can't  enforce deny mode locking cluster-wide since it's not supported there in any fashion.

FWIW, POSIX doesn't really have any sort of share/deny mode locking (though apparently BSD has O_SHLOCK and O_EXLOCK, which is close). NFS and SMB both do however, and need to enforce those semantics. So, we need this behavior in them at least, or we can't handle the case of a cached deny mode lock on the client. For other callers, it not as crucial:

Suppose we have a server application (like ganesha or samba) that has a deny lock set such that another client would be denied an open, and then a "normal" ceph client (e.g. ceph-fuse or whatever) comes along and opens the file?

He'll get an open and eventually be able to do I/O to the file, like normal. This is primarily to ensure that we can at least enforce it within a single application (ganesha or samba). Eventually we may need to add something for cluster-coherent deny locks, but I'm not proposing any mechanism like that here.

jtlayton commented on src/client/Inode.cc:
```diff
@@ -549,3 +557,175 @@ void Inode::set_async_err(int r)
+    return false;
+
+  Delegation& deleg = delegations.front();
+  return deleg.is_recalled();
```
Exactly. I'll try to flesh out the comments a bit there.

jtlayton commented on src/client/Inode.cc:
```diff
@@ -549,3 +557,175 @@ void Inode::set_async_err(int r)
+  }
+}
+
+bool Inode::delegations_broken(bool skip_read)
```
Fair enough. That should be doable.

jtlayton commented on src/client/Inode.cc:
```diff
@@ -549,3 +557,175 @@ void Inode::set_async_err(int r)
   }
 }
 
+class C_C_Deleg_Timeout : public Context {
```
Heh, I cargo culted that naming from C_C_Tick which existed in the tree until recently. I'm fine with renaming it. Maybe InodeDelegTimeout then?

gregsfortytwo replied:
Ah, I think the C_C_Tick was just a safer C_Tick or something, is why it got named that silly way. :)

jtlayton commented on src/client/Inode.cc:
```diff
@@ -549,3 +557,175 @@ void Inode::set_async_err(int r)
+    Context *timeout_event = new C_C_Deleg_Timeout(&deleg);
+    deleg.recall(skip_read);
+    deleg.arm_timeout(&client->timer, timeout_event,
+		      client->get_deleg_timeout());
```
Ouch, good catch. Yes, we don't want to arm the timeout if we're skipping a read delegation. Will fix.

jtlayton commented on src/client/Delegation.cc:
```diff
@@ -0,0 +1,115 @@
+Delegation::~Delegation()
+{
+  Inode *inode = fh->inode.get();
+  inode->client->put_cap_ref(inode, ceph_deleg_caps_for_type(type));
```
Hmm...the problem there is that I'm relying on the Inode routines to pass in the SafeTimer to use.

In practice the destructor will always be called after the timer is disarmed, either via an explicit CEPH_DELEGATION_NONE request, or the timer popping and unmounting the whole cmount.

jtlayton commented on src/client/Delegation.cc:
```diff
@@ -0,0 +1,115 @@
+
+void Delegation::reinit(unsigned _type, ceph_deleg_cb_t _recall_cb, void *_priv)
+{
+  Inode *inode = fh->inode.get();
```
This is just assigning to a local variable that we use later. It should probably be scoped inside that if block though. I'll clean that up.

gregsfortytwo replied:
Oh yeah, I incorrectly thought fh was one of the parameters being passed in here for some reason. :o

jtlayton commented on src/client/Delegation.cc:
```diff
@@ -0,0 +1,115 @@
+  if (timeout_event) {
+    delete event;
+    return;
+  }
```
Good point. I may move the event allocation into the Delegation class routines. Let me think on that a bit.

jtlayton commented:
A small set of updates that should make this a bit cleaner.

gregsfortytwo commented:
Updates look better 👍
Just worried about those cluster-wide behavior things, which may not actually be a problem...?

jtlayton commented:
Thanks. It's a fair question and a bit of a subtle point. Essentially we don't want to change other callers. This tunable exists to allow a single application to enforce those semantics properly on top of cephfs which doesn't yet support it.

jtlayton commented:
I think this should cover the last of @gregsfortytwo 's review. This ensures that we always cap the delegation timeout to a value 1s less than the session timeout. Ideally, the session timeout would be higher by default.

Typical lease/delegation timeouts are 90s, it might be nice to consider doubling the default session timeout to 120s (or recommending that when setting up ganesha on cephfs).

gregsfortytwo commented:
Okay, I'm down with not changing the other client cap behavior. I tend to assume we want to be more correct rather than less, in cases where that's not too difficult, but I guess that's not really a sensible statement to make regarding deny-mode locks in a case where we'll still force back caps and service the open in any case.

jtlayton commented:
retest this please

jtlayton commented:
The important bit here is that this doesn't preclude us from adding "real" deny mode lock support in the future, if we decide that it's worthwhile for samba and ganesha. For now, we allow other sorts of clients to get the behavior  to which they are accustomed.

jtlayton commented:
For now, I'll go ahead and close this review. I'll plan to reopen it once I've squashed down the patches and implemented @gregsfortytwo's suggestion for handling the timeouts.

jtlayton commented:
For now, I'll go ahead and close this review. I'll plan to reopen it once I've squashed down the patches and implemented @gregsfortytwo's suggestion for handling the timeouts.
